crowbar_framework/app/models/proposal_object.rb

```diff
@@ -155,106 +121,36 @@ def barclamp
-    else
-      return "unready" if bc["crowbar-committing"]
-      return "pending" if bc["crowbar-queued"]
-      return "hold" if !bc.has_key? "crowbar-queued" and !bc.has_key? "crowbar-committing"
-      if !bc.key? "crowbar-status" or bc["crowbar-status"] === "success"
-        "ready"
-      else
-        "failed"
-      end
-    end
-  end
-  
-  # nil if not appliciable, true = if success, false if failed
-  def failed?
-     status === 'failed'
+  def save(options = {})
+    self.latest_applied = !!options[:applied]
+    increment_crowbar_revision!
+    Rails.logger.debug("Saving data bag item: #{@item["id"]} - #{crowbar_revision}")
+    @item.save
+    save_proposal_in_sqlite
+    Rails.logger.debug("Done saving data bag item: #{@item["id"]} - #{crowbar_revision}")
   end
 
-  # for locationlization, will lookup text before the :  
-  def fail_reason
-     s = if failed?
-       @item["deployment"][self.barclamp]["crowbar-failed"].to_s
-     elsif status === "ready"
-       "Did not fail.  Successfully applied: #{barclamp}-#{name} (status #{status})"
-     else
-       "No success information for proposal: #{barclamp}-#{name} (status #{status})"
-     end
-     out = s.split(":")
-     out[0] = I18n.t out[0], :default=> out[0]
-     return out.join(":").to_s
-  end
-  
-  def description
-    @item['description']
+  def destroy
+    Rails.logger.debug("Destroying data bag item: #{@item["id"]} - #{crowbar_revision}")
+    @item.destroy(@item.data_bag, @item["id"])
+    delete_proposal_from_sqlite
+    Rails.logger.debug("Done removal of data bag item: #{@item["id"]} - #{crowbar_revision}")
   end
   
-  def elements
-    @item.raw_data['deployment'][self.barclamp]["elements"]
-  end
-
-  def all_elements
-    @item.raw_data['deployment'][self.barclamp]["element_order"].flatten.uniq
-  end
-
-  def role
-    RoleObject.find_role_by_name("#{barclamp}-config-#{name}")
-  end
-
-  def crowbar_revision
-    @item["deployment"][barclamp]["crowbar-revision"].to_i rescue 0
-  end
-
-  def latest_applied?
-    @item["deployment"][barclamp]["crowbar-applied"] rescue false
-  end
-
-  def latest_applied=(applied)
-    @item["deployment"] ||= {}
-    @item["deployment"][barclamp] ||= {}
-    @item["deployment"][barclamp]["crowbar-applied"] = applied
-  end
-
-  def active?
-    !role.nil?
-  end
-
-  def raw_data
-    @item.raw_data
-  end
+  private
 
-  def raw_data=(value)
-    @item.raw_data = value
-  end
+  def delete_proposal_from_sqlite
+    attrs = { barclamp: self.barclamp, name: self.name }
 
-  def [](attrib)
-    @item[attrib]
+    prop = Proposal.where(attrs).first
+    prop.destroy if prop
   end
 
-  def []=(attrib, value)
-    @item[attrib] = value
-  end
+  def save_proposal_in_sqlite
+    attrs = { barclamp: self.barclamp, name: self.name }
 
-  def initialize(x)
-    @item = x
+    prop = Proposal.where(attrs).first_or_initialize(attrs)
+    prop.update(attrs.merge(properties: @item.raw_data)) if @item.raw_data != prop.raw_data
```

Comment:
> Will the `@item.raw_data != prop.raw_data` compare strings, or JSON/Hashes? If the former, then I think it's not correct as the string can change without the JSON changing.

It should compare hashes, I covered this by a test here:

https://github.com/crowbar/barclamp-crowbar/pull/1254/files#diff-f125db84446056a04e9a58bb6052128cR45

or maybe you see a different scenario, where this would fail?